src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -106,12 +110,15 @@ public int Count
             }
         }
 
-        public Enumerator GetEnumerator()
+        public T this[int index]
         {
-            if (_array == null)
-                ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
-            Contract.EndContractBlock();
+            get { return _array[_offset + index]; }
```

Comment:
Do you have any example where ArraySegment is used in perf-sensitive place?

`ArraySegment<T>` is in the same category as `List<T>` in my mind - pretty convenient and lightweight, but still adds some overhead. I do not think that the perf-sensitive algorithms should be written against `ArraySegment<T>`. They should be written directly against array (or potentially `Span<T>` in future).